### PR TITLE
Fix race condition in Task-based Socket wrappers

### PR DESCRIPTION
**Pull request by stephentoub**
My recent optimization to put the Task-based Socket send/receive operations on top of the SocketAsyncEventArgs methods instead of on top of the APM methods introduced a race condition that's causing some tests to hang.  If a send/receive is initiated, completes asynchronously, but completes asynchronously so fast that it happens before the caller has progressed past the call, the asynchronous completion callback may return the SAEA to the pool (zero'ing out the builder) before the caller gets a chance to access the builder.  At that point, when the caller does access the builder, it'll effectively be a different builder, and bad things ensue (like waiting for a completion notification that'll never happen because it already happened and completed a different builder).

This is the cause of the recent rash of build timeouts in CI, in particular on Unix, and in particular stemming from the System.IO.Pipes tests (named pipes on Unix are implemented on sockets).

This fixes it by ensuring that the right party gets the responsibility for returning the SAEA to the pool only once both the caller and the callback have accessed it and gotten from it what they need.

cc: @geoffkizer

### CODE REVIEW

**Comment by stephentoub** [comment]
@dotnet-bot test Outerloop Windows_NT Debug please
@dotnet-bot test Outerloop Ubuntu14.04 Debug please
@dotnet-bot test Outerloop OSX Debug please

**Comment by stephentoub** [comment]
I'm going to merge this to unblock CI.  Will address any PR feedback separately.

File: src/System.Net.Sockets/src/System/Net/Sockets/Socket.Tasks.cs [hunk 1]
```diff
@@ -140,25 +140,36 @@ internal Task<int> ReceiveAsync(ArraySegment<byte> buffer, SocketFlags socketFla
             if (saea.BufferList != null) saea.BufferList = null;
             saea.SetBuffer(buffer.Array, buffer.Offset, buffer.Count);
             saea.SocketFlags = socketFlags;
-            saea.WrapExceptionsInIOExceptions = wrapExceptionsInIOExceptions;
+            saea._wrapExceptionsInIOExceptions = wrapExceptionsInIOExceptions;
 
             // Initiate the receive
             Task<int> t;
+            bool responsibleForReturningToPool;
```
Comment (geoffkizer) [inline src/System.Net.Sockets/src/System/Net/Sockets/Socket.Tasks.cs]: Re this pattern, I think it's slightly clearer to encapsulate the logic for sync vs async completion, e.g.:

```
if (!ReceiveAsync(saea))
{
    t = GetSyncCompletionTask(saea);
}
else
{
    t = GetPendingTask(saea);
}
```

Or even just

```
t = GetTask(saea, ReceiveAsync(saea));
```

Thoughts?

Comment (stephentoub) [inline src/System.Net.Sockets/src/System/Net/Sockets/Socket.Tasks.cs]: Makes sense, I'll take a look at encapsulating it tomorrow.

**Pull request merged**